Exercises/chapter-11/e3-promise-all.js

```diff
@@ -0,0 +1,47 @@
+function Promise_all(promises) {
+    return new Promise((resolve, reject) => {
+        const results = [];
+        try {
+            if (promises.length === 0) {
+                resolve(results);
+                return;
+            }
+
+            for (let i = 0; i < promises.length; i++) {
+                promises[i].then(data => {
+                    results[i] = data;
+                    if (i === promises.length-1) {
+                        resolve(results);
+                    }
+                }).catch(reject);
+            }
+        } catch (e) {
+            reject;
+        }
+    });
+}
+  
+// Test code.
+Promise_all([]).then(array => {
+    console.log("This should be []:", array);
+});
+
+function soon(val) {
+    return new Promise(resolve => {
+        setTimeout(() => resolve(val), Math.random() * 500);
+    });
+}
+
+Promise_all([soon(1), soon(2), soon(3)]).then(array => {
```

Comment:
To test the race conditions you should shorten soon and run a bunch of iterations of this.